Fixes for #906 (TypeError with many=True) and #907 (return [] not {} on many=True error) and #930 (disallows strings and dicts on many=True)

tuukkamustonen opened the pull request:
See #906 and #907. In same PR, as latter one re-uses fix and test of the former.

Do you want me to move the test from `test_schema.py` to `test_marshalling.py` (as the fixes are in `marshalling` module?

Also, I wonder if these need to be reflected in `Marshaller.serialize()`, too? What is the intended behavior if you try to marshall a primitive, for example?

I took liberty to thank myself in changelog :)

tuukkamustonen committed:
Fix #906: Raise `ValidationError` (not `TypeError`) on non-iterables with `many=True`

tuukkamustonen committed:
Fix #907: Return empty list as `ValidationError.valid_data` with many=True

Only when there's no valid data!

tuukkamustonen commented on marshmallow/marshalling.py:
```diff
@@ -190,7 +190,7 @@ def run_validator(
     def deserialize(
         self, data, fields_dict, many=False, partial=False,
-        unknown=RAISE, dict_class=dict, index_errors=True, index=None,
+        unknown=RAISE, dict_class=dict, list_class=list, index_errors=True, index=None,
```
Not sure about the use case, but `dict_class` is configurable, so...

sloria replied:
I don't think `list_class` is necessary. The only reason we have `dict_class` is to support the `ordered = True` option (in which case `OrderedDict` is used instead of `dict`).

tuukkamustonen replied:
Alright, I'll remove the option.

tuukkamustonen commented on marshmallow/marshalling.py:
```diff
@@ -230,12 +232,10 @@ def deserialize(
-        ret = dict_class()
-        # Check data is a dict
         if not isinstance(data, collections.Mapping):
             self.store_error(SCHEMA, ('Invalid input type.', ), index=index)
```
The error for non-iterables is logged here. I don't like the flow of code now - it's somewhat confusing - but I didn't want to nest the code to leave minimal diff.

tuukkamustonen replied:
Oh, and I moved `partial_is_collection` slightly, because it's not needed on the main level. I can restore original code if you want.

sloria replied:
> The error for non-iterables is logged here. I don't like the flow of code now - it's somewhat confusing - but I didn't want to nest the code to leave minimal diff.

Fair enough. Feel free to make further improvements to the flow here in a future PR.

sloria commented:
Thanks for the PR. I'll try to review this sometime over the weekend.

sloria commented on marshmallow/marshalling.py:
```diff
@@ -203,24 +203,26 @@ def deserialize(
         """
-        if many and data is not None:
+        ret = dict_class() if not many else list_class()
+        if many and data is not None and is_iterable(data):
```
Why not use `is_iterable_but_not_string` here?

tuukkamustonen replied:
In my understanding, the library regards strings as iterable (which they are) so I didn't want to make an exception here. If you pass string with many=True, marshmallow will iterate over string and pick individual characters as elements, so I think we should act the same way here.

(However, I think it's confusing and weird, because you cannot validate "primitives" with marshmallow anyway (afaik), and accidentally passing in string rather than "actual" iterable such as list should rather raise than pass, but that's just my perspective. I actually have a PR on this in https://github.com/marshmallow-code/marshmallow/pull/622 but there was no clear outcome.)

sloria replied:
> However, I think it's confusing and weird, because you cannot validate "primitives" with marshmallow anyway (afaik), and accidentally passing in string rather than "actual" iterable such as list should rather raise than pass, but that's just my perspective

Exactly. Passing in a string is invalid usage, so I think raising an error is the correct behavior.

tuukkamustonen replied:
> Exactly. Passing in a string is invalid usage, so I think raising an error is the correct behavior.

But marshmallow doesn't currently give error when you pass string as data for `many=True`? Or 2.x at least didn't, was it changed for 3.x?

sloria replied:
I don't believe 2.x raises an error, but I don't think that should be dispositive. 3.x became more strict in its handling of `None`, so I think it should do the same for strings (and other invalid types).

tuukkamustonen replied:
Ok, I'll check that too tomorrow. In a way regarding strings like any iterable makes sense, but it's normally not what the user intended so raising error makes more sense. So we agree on that, nice.

tuukkamustonen replied:
@sloria Checked this, marshmallow 3.x still considers a `str` valid for `many=True`:

```python
In [1]: from marshmallow import Schema

In [3]: from marshmallow import fields

In [4]: class Sch(Schema):
   ...:     foo = fields.Int()

In [7]: Sch(many=True).validate([{'foo': 1}, {'foo': 2}])
Out[7]: {}

In [8]: Sch(many=True).validate('asf')
Out[8]: 
{0: {u'_schema': [u'Invalid input type.']},
 1: {u'_schema': [u'Invalid input type.']},
 2: {u'_schema': [u'Invalid input type.']}}
```

Rather than giving out error `{'_schema': ['Invalid input type.']}`.

Could you check #622 (the discussion there, basically). What is your take on it?

sloria commented on marshmallow/utils.py:
```diff
@@ -52,6 +52,11 @@ def is_generator(obj):
     return inspect.isgeneratorfunction(obj) or inspect.isgenerator(obj)
 
 
+def is_iterable(obj):
```
See previous comment: not sure this is necessary--I think you can just use `is_iterable_but_not_string`.

tuukkamustonen replied:
Will remove if it's not needed.

sloria reviewed (changes requested):
Thanks for the PR! I appreciate that you updated the changelog.

Just a few nits to address; when those are done, this is good to merge.

tuukkamustonen commented:
@sloria Made the adjustments:

- Removed `list_class` option
- Changed to `is_iterable_but_not_string()`

The latter adjustment CHANGES behavior in that passing `str` to `many=True` will now raise `{'_schema': ['Invalid input type.']}`. I think we should write a changelog entry also on this...

tuukkamustonen committed:
Fix #930: Raise `ValidationError` on strings and mappings with `many=True`

The actual fix actually came together in #906, just expanding it to cover also
dicts, plus adding tests and docs here.

tuukkamustonen commented:
@sloria I added a commit that tests `many=True` and a CHANGELOG entry (new commit). Also opened #930 for it. If we now would block strings, should we also reject dicts?

sloria commented:
@tuukkamustonen Hm, yes, I suppose `dicts` should be dis-allowed, too.

tuukkamustonen commented:
Ok, added `collections.Mapping` exception (instead of raw `dict`). Squashed into previous commit.

For consistence, should we tackle strings and mappings also on `dump()`? Trying to dump string with `many=True` now gives `AttributeError` and dumping dict just works. I guess `dump()` is so different beast we don't need to touch it?

tuukkamustonen commented:
Resolved conflicts.

tuukkamustonen commented:
@sloria @lafrech Should we try to push it in? Are there any adjustments still that you want me to apply?

sloria commented:
This is looking good. Since these are bug fixes, though, I wonder if they should be sent to `2.x-line`. @lafrech  Thoughts?

lafrech commented on marshmallow/marshalling.py:
```diff
@@ -203,24 +203,25 @@ def deserialize(
         :param unknown: Whether to exclude, include, or raise an error for unknown
             fields in the data. Use `EXCLUDE`, `INCLUDE` or `RAISE`.
-        :param type dict_class: Dictionary class used to construct the output.
+        :param type dict_class: Dictionary class used to construct the output when many=False.
```
The class is used also when many=True, to construct the output of each element. I'd remove this change.

lafrech commented on marshmallow/marshalling.py:
```diff
@@ -203,24 +203,25 @@ def deserialize(
         """
-        if many and data is not None:
+        ret = dict_class() if not many else []
+        if many and data is not None and is_collection(data):
```
`data is not None` is useless, here, as `is_collection(None)` is `False`.

lafrech commented:
Although I had worked on this code already, I found the logic hard to grab at first sight.

```py
if many and data is not None and is_collection(data):
    [...]  # Normal many=True case
if not isinstance(data, collections.Mapping) or many:
    [...]  # (many=False and not dict) or (many=True and not collection) case
else:
    [...]  # Normal many=False case
```
I reworked it in https://github.com/marshmallow-code/marshmallow/pull/954 while also addressing my remarks above.

My version duplicates `self.store_error(SCHEMA, ('Invalid input type.', ), index=index)` but I don't think it is that band and overall I find it easier to understand.

No strong opinion, @sloria, pick the one you prefer.

> Since these are bug fixes, though, I wonder if they should be sent to 2.x-line. 

I don't think anyone would [reasonably](https://xkcd.com/1172/) rely on current behaviour, so fixing this in 2.x makes sense. How would you proceed? Would you apply the same patches manually? Cherry-pick the commits? Not sure which is easier.

sloria commented:
Thanks @lafrech. I agree with your changes on that PR. Let's go with that.

> How would you proceed? Would you apply the same patches manually? Cherry-pick the commits? Not sure which is easier.

I can take care of cherry-picking the commits into a branch off of `2.x-line`.

Closing. Let's continue in #954.
